### PR TITLE
CBL-3698: Repeat the tests in ReplicatorCollectionTest.cc with Sync-g…

### PR DESCRIPTION
**Pull request by jianminzhao**
…ateway server.

Tests in ReplicatorCollectionTest.cc are performed with loopback connections between two replicators in the same executable. We port those tests with Sync-gateway server on the local machine, as appropriate to test the interaction between the replicator in CBL as client and Sync-gateway as the server. The new tests are put in ReplicatorCollectionSGTest.cc. Still to be done is "Resolve Conflict", "Reset Checkpoint with Pull", and "Filters & docIDs with Multiple Collections".

### CODE REVIEW

**Review by borrrden — COMMENTED** [review]
This PR is too huge to reasonably review, as it does far more than simply add tests for SG it seems.  It changes a whole bunch of functions that every other test also uses, and for what?

**Comment by jianminzhao** [comment]
Perhaps it's candidate of in-person live review.

File: Replicator/tests/ReplicatorCollectionSGTest.cc [hunk 1]
```diff
@@ -0,0 +1,1094 @@
...
+ --header "Authorization: Basic QWRtaW5pc3RyYXRvcjpwYXNzd29yZA==" \
+ --data-raw '{"num_index_replicas": 0, "bucket": "your_bucket_name", "scopes": {"flowers": {"collections":{"flowers":{}}}}}'
+ */
+//  config SG user:
+/*
+ curl --location --request POST "localhost:4985/scratch/_user/" \
+ --header "Content-Type: application/json" \
+ --header "Authorization: Basic QWRtaW5pc3RyYXRvcjpwYXNzd29yZA==" \
+ --data-raw '{"name": "sguser", "password": "password", "admin_channels": ["*"]}'
+ */
```
Comment (pasin) [inline Replicator/tests/ReplicatorCollectionSGTest.cc:68]: * Would it be possible to create a static SG config file to configure SG? Previous test setup has two config files : TLS and Non-TLS.

* Would it be possible to have a script that helps running the docker and also configure the buckets?

Comment (jianminzhao) [reply Replicator/tests/ReplicatorCollectionSGTest.cc:68]: 1. I got the impression that static SG config is obsolete, contraindicated
2. yes. we may provide the script for Unix shell.

Comment (pasin) [reply Replicator/tests/ReplicatorCollectionSGTest.cc:68]: Here is a bit more info from Adam:

```
It's correct to say that we won't support file-based database configuration forever, and the recommended
approach for 3.0 and later is to use file-based config only for cluster bootstrapping, and to use the REST API
for database and user creation/management
```

Comment (pasin) [reply Replicator/tests/ReplicatorCollectionSGTest.cc:68]: As the config for tests are pretty much static, we could move the CURL commands to the same script that starts SG. If there are some changes in the test code, the test code will need to do the REST API call.

File: Replicator/tests/ReplicatorCollectionSGTest.cc [hunk 2]
```diff
@@ -0,0 +1,1094 @@
...
+        c4Params.collections = replCollections.data();
+    };
+    replicate(paramsSetter);
+    verifyDocs(collectionSpecs, docIDs);
+    for (size_t i = 0; i < collectionCount; ++i) {
+        checkAttachments(verifyDb, blobKeys[i], attachments1);
+    }
+}
+
+#if 0
```
Comment (pasin) [inline Replicator/tests/ReplicatorCollectionSGTest.cc:683]: Is this test failed? Can you give info and ticket number if having one about the reason to disable this test?

File: Replicator/tests/ReplicatorCollectionSGTest.cc [hunk 3]
```diff
@@ -0,0 +1,1094 @@
...
+static constexpr C4CollectionSpec Guitars = { GuitarsName, kC4DefaultScopeID };
+
+static constexpr slice RosesName = "roses"_sl;
+static constexpr slice TulipsName = "tulips"_sl;
+//static constexpr slice LavenderName = "lavenders"_sl;
+static constexpr slice FlowersScopeName = "flowers"_sl;
+
+static constexpr C4CollectionSpec Roses = { RosesName, FlowersScopeName };
+static constexpr C4CollectionSpec Tulips = { TulipsName, FlowersScopeName };
+//static constexpr C4CollectionSpec Lavenders = { LavenderName, FlowersScopeName };
```
Comment (pasin) [inline Replicator/tests/ReplicatorCollectionSGTest.cc:84]: Remove if not needed.

**Review by pasin — CHANGES_REQUESTED** [review]

File: Replicator/tests/ReplicatorCollectionSGTest.cc (hunk 1 above)
Comment (jianminzhao) [reply Replicator/tests/ReplicatorCollectionSGTest.cc:68]: I haven't checked in any script yet. I hope we will script to do the SG and CB server. Jim has script in his docker compose. I just did not get to it and manually set up the cb server. I hope not to include it in this PR if okay.

File: Replicator/tests/ReplicatorCollectionSGTest.cc [hunk 4]
```diff
@@ -0,0 +1,1094 @@
...
+        thread.join();
+    }
+};
+
+TEST_CASE_METHOD(ReplicatorCollectionSGTest, "Multiple Collections Incremental Revisions SG", "[.SyncServerCollection]") {
+    string idPrefix = timePrefix();
+    constexpr size_t collectionCount = 1;
+    std::array<C4CollectionSpec, collectionCount> collectionSpecs = {
+        Roses
+        //, Tulips
```
[the remark this answers is not in the document]
Comment (borrrden) [reply Replicator/tests/ReplicatorCollectionSGTest.cc:585]: The estimate for that is the end of October

**Review by pasin — APPROVED** [review]

**Pull request merged**